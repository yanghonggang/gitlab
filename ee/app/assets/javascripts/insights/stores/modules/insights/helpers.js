import { CHART_TYPES } from 'ee/insights/constants';
import { __ } from '~/locale';

const getAxisTitle = label => {
  switch (label) {
    case 'day':
      return __('Days');
    case 'week':
      return __('Weeks');
    case 'month':
      return __('Months');
    case 'issue':
      return __('Issues');
    case 'merge_request':
      return __('Merge requests');
    default:
      return '';
  }
};

export const transformChartDataForGlCharts = (
  { type, query: { group_by, issuable_type } },
  { labels, datasets },
) => {
  const formattedData = {
    xAxisTitle: getAxisTitle(group_by),
    yAxisTitle: getAxisTitle(issuable_type),
    labels,
    datasets: [],
    seriesNames: [],
  };

  switch (type) {
    case CHART_TYPES.BAR:
      formattedData.datasets = [
        {
          name: 'all',
          data: labels.map((label, i) => [label, datasets[0].data[i]]),
        },
      ];
      break;
    case CHART_TYPES.STACKED_BAR:
      formattedData.datasets.push(
        ...datasets.map(dataset => ({
          name: dataset.label,
          data: dataset.data,
        })),
      );
      break;
    case CHART_TYPES.LINE:
      formattedData.datasets.push(
        ...datasets.map(dataset => ({
          name: dataset.label,
          data: labels.map((label, i) => [label, dataset.data[i]]),
        })),
      );

      break;
    default:
      formattedData.datasets = { all: labels.map((label, i) => [label, datasets[0].data[i]]) };
  }
  return formattedData;
};

export default {
  transformChartDataForGlCharts,
};
