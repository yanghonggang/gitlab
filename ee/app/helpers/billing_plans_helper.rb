# frozen_string_literal: true

module BillingPlansHelper
  def subscription_plan_info(plans_data, current_plan_code)
    plans_data.find { |plan| plan.code == current_plan_code }
  end

  def number_to_plan_currency(value)
    number_to_currency(value, unit: '$', strip_insignificant_zeros: true, format: "%u%n")
  end

  def subscription_plan_data_attributes(group, plan)
    return {} unless group

    {
      namespace_id: group.id,
      namespace_name: group.name,
      plan_upgrade_href: plan_upgrade_url(group, plan),
      customer_portal_url: "#{EE::SUBSCRIPTIONS_URL}/subscriptions"
    }
  end

  def use_new_purchase_flow?(namespace)
    namespace.group? &&
      namespace.actual_plan_name == Plan::FREE
  end

  def show_contact_sales_button?(purchase_link_action)
    experiment_enabled?(:contact_sales_btn_in_app) &&
      purchase_link_action == 'upgrade'
  end

  def experiment_tracking_data_for_button_click(button_label)
    return {} unless Gitlab::Experimentation.enabled?(:contact_sales_btn_in_app)

    {
      track: {
        event: 'click_button',
        label: button_label,
        property: experiment_tracking_category_and_group(:contact_sales_btn_in_app)
      }
    }
  end

  def plan_feature_short_list(plan)
    return [] unless plan.features

    plan.features.sort_by! { |feature| feature.highlight ? 0 : 1 }[0...4]
  end

  def plan_purchase_or_upgrade_url(group, plan, current_plan)
    if group.upgradable?
      plan_upgrade_url(group, current_plan)
    else
      plan_purchase_url(group, plan)
    end
  end

  def show_plans?(namespace)
    namespace.trial_active? || !namespace.gold_plan?
  end

  def show_trial_banner?(namespace)
    return false unless params[:trial]

    root = namespace.has_parent? ? namespace.root_ancestor : namespace
    root.trial_active?
  end

  def namespace_for_user?(namespace)
    namespace == current_user.namespace
  end

  def seats_data_last_update_info
    last_enqueue_time = UpdateMaxSeatsUsedForGitlabComSubscriptionsWorker.last_enqueue_time&.utc
    return _("Seats usage data as of %{last_enqueue_time} (Updated daily)" % { last_enqueue_time: last_enqueue_time }) if last_enqueue_time

    _('Seats usage data is updated every day at 12:00pm UTC')
  end

  private

  def plan_purchase_url(group, plan)
    if use_new_purchase_flow?(group)
      new_subscriptions_path(plan_id: plan.id, namespace_id: group.id)
    else
      "#{plan.purchase_link.href}&gl_namespace_id=#{group.id}"
    end
  end

  def plan_upgrade_url(group, plan)
    return unless group && plan&.id

    "#{EE::SUBSCRIPTIONS_URL}/gitlab/namespaces/#{group.id}/upgrade/#{plan.id}"
  end
end
