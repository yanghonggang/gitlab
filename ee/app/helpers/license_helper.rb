# frozen_string_literal: true

module LicenseHelper
  include ActionView::Helpers::AssetTagHelper
  include ActionView::Helpers::UrlHelper

  delegate :new_admin_license_path, to: 'Gitlab::Routing.url_helpers'

  def seats_calculation_message(license)
    return unless license.exclude_guests_from_active_count?

    s_("Users with a Guest role or those who don't belong to a Project or Group will not use a seat from your license.")
  end

  def current_license_title
    License.current&.plan&.titleize || 'Core'
  end

  def new_trial_url
    return_to_url = CGI.escape(Gitlab.config.gitlab.url)
    uri = URI.parse(::EE::SUBSCRIPTIONS_URL)
    uri.path = '/trials/new'
    uri.query = "return_to=#{return_to_url}&id=#{Base64.strict_encode64(current_user.email)}"
    uri.to_s
  end

  def show_promotions?(selected_user = current_user)
    return false unless selected_user

    if Gitlab::CurrentSettings.current_application_settings
      .should_check_namespace_plan?
      true
    else
      license = License.current
      license.nil? || license.expired?
    end
  end

  def show_advanced_search_promotion?
    !Gitlab::CurrentSettings.should_check_namespace_plan? && show_promotions? && show_callout?('promote_advanced_search_dismissed') && !License.feature_available?(:elastic_search)
  end

  def licensed_users(license)
    if license.restricted?(:active_user_count)
      number_with_delimiter(license.restrictions[:active_user_count])
    else
      _('Unlimited')
    end
  end

  extend self
end
