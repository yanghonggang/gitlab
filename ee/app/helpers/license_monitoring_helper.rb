# frozen_string_literal: true

module LicenseMonitoringHelper
  include Gitlab::Utils::StrongMemoize

  def show_active_user_count_threshold_banner?
    return if ::Gitlab.com?
    return unless admin_section?
    return if user_dismissed?(UserCalloutsHelper::ACTIVE_USER_COUNT_THRESHOLD)
    return if license_not_available_or_trial?

    current_user&.admin? && current_license.active_user_count_threshold_reached?
  end

  def users_over_license
    strong_memoize(:users_over_license) do
      license_is_over_capacity? ? current_license_overage : 0
    end
  end

  private

  def license_is_over_capacity?
    return if ::Gitlab.com?
    return if license_not_available_or_trial?

    current_license_overage > 0
  end

  def license_not_available_or_trial?
    current_license.nil? || current_license.trial?
  end

  def current_license
    strong_memoize(:current_license) { License.current }
  end

  def current_license_overage
    strong_memoize(:current_license_overage) { current_license.overage_with_historical_max }
  end

  def active_user_count_threshold
    strong_memoize(:active_user_count_threshold) { current_license.active_user_count_threshold }
  end

  def total_user_count
    strong_memoize(:total_user_count) { current_license.restricted_user_count }
  end

  def remaining_user_count
    strong_memoize(:remaining_user_count) { current_license.remaining_user_count }
  end
end
