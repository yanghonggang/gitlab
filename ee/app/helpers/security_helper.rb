# frozen_string_literal: true

module SecurityHelper
  def instance_security_dashboard_data
    {
      dashboard_documentation: help_page_path('user/application_security/security_dashboard/index', anchor: 'instance-security-dashboard'),
      no_vulnerabilities_svg_path: image_path('illustrations/issues.svg'),
      empty_dashboard_state_svg_path: image_path('illustrations/security-dashboard-empty-state.svg'),
      empty_state_svg_path: image_path('illustrations/operations-dashboard_empty.svg'),
      project_add_endpoint: security_projects_path,
      project_list_endpoint: security_projects_path,
      instance_dashboard_settings_path: settings_security_dashboard_path,
      vulnerabilities_export_endpoint: expose_path(api_v4_security_vulnerability_exports_path)
    }
  end

  def security_dashboard_unavailable_view_data
    {
      empty_state_svg_path: image_path('illustrations/security-dashboard-empty-state.svg'),
      dashboard_documentation: help_page_path('user/application_security/security_dashboard/index'),
      is_unavailable: "true"
    }
  end
end
