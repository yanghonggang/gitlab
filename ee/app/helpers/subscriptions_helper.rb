# frozen_string_literal: true

module SubscriptionsHelper
  include ::Gitlab::Utils::StrongMemoize

  def subscription_data
    {
      setup_for_company: (current_user.setup_for_company == true).to_s,
      full_name: current_user.name,
      plan_data: plan_data.to_json,
      plan_id: params[:plan_id],
      namespace_id: params[:namespace_id],
      new_user: new_user?.to_s,
      group_data: group_data.to_json
    }
  end

  def plan_title
    strong_memoize(:plan_title) do
      plan = plan_data.find { |plan| plan[:id] == params[:plan_id] }
      plan[:code].titleize if plan
    end
  end

  private

  def new_user?
    return false unless request.referer.present?

    URI.parse(request.referer).path == users_sign_up_welcome_path
  end

  def plan_data
    FetchSubscriptionPlansService.new(plan: :free).execute
      .map(&:symbolize_keys)
      .reject { |plan| plan[:free] }
      .map { |plan| plan.slice(:id, :code, :price_per_year) }
  end

  def group_data
    current_user.manageable_groups_eligible_for_subscription.with_counts(archived: false).map do |namespace|
      {
        id: namespace.id,
        name: namespace.name,
        users: namespace.member_count
      }
    end
  end
end
