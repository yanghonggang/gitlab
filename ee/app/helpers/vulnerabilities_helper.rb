# frozen_string_literal: true

module VulnerabilitiesHelper
  def vulnerability_details_json(vulnerability, pipeline)
    vulnerability_details(vulnerability, pipeline).to_json
  end

  def vulnerability_details(vulnerability, pipeline)
    return unless vulnerability

    result = {
      timestamp: Time.now.to_i,
      create_issue_url: create_issue_url_for(vulnerability),
      has_mr: !!vulnerability.finding.merge_request_feedback.try(:merge_request_iid),
      create_mr_url: create_vulnerability_feedback_merge_request_path(vulnerability.finding.project),
      discussions_url: discussions_project_security_vulnerability_path(vulnerability.project, vulnerability),
      notes_url: project_security_vulnerability_notes_path(vulnerability.project, vulnerability),
      vulnerability_feedback_help_path: help_page_path('user/application_security/index', anchor: 'interacting-with-the-vulnerabilities'),
      related_issues_help_path: help_page_path('user/application_security/index', anchor: 'managing-related-issues-for-a-vulnerability'),
      pipeline: vulnerability_pipeline_data(pipeline),
      can_modify_related_issues: current_user.can?(:admin_vulnerability_issue_link, vulnerability),
      issue_tracking_help_path: help_page_path('user/project/settings', anchor: 'sharing-and-permissions'),
      permissions_help_path: help_page_path('user/permissions', anchor: 'project-members-permissions')
    }

    result.merge(vulnerability_data(vulnerability), vulnerability_finding_data(vulnerability))
  end

  def create_issue_url_for(vulnerability)
    return unless vulnerability.project.issues_enabled?

    create_issue_project_security_vulnerability_path(vulnerability.project, vulnerability)
  end

  def vulnerability_pipeline_data(pipeline)
    return unless pipeline

    {
      id: pipeline.id,
      created_at: pipeline.created_at.iso8601,
      url: pipeline_path(pipeline),
      source_branch: pipeline.ref
    }
  end

  def vulnerability_data(vulnerability)
    VulnerabilitySerializer.new.represent(vulnerability)
  end

  def vulnerability_finding_data(vulnerability)
    finding = Vulnerabilities::FindingSerializer.new(current_user: current_user).represent(vulnerability.finding)

    data = finding.slice(
      :description,
      :identifiers,
      :links,
      :location,
      :name,
      :issue_feedback,
      :merge_request_feedback,
      :project,
      :project_fingerprint,
      :remediations,
      :evidence,
      :scanner,
      :solution,
      :request,
      :response,
      :evidence_source,
      :supporting_messages,
      :assets
    )

    if data[:location]['file']
      branch = vulnerability.finding.pipelines&.last&.sha || vulnerability.project.default_branch
      path = project_blob_path(vulnerability.project, tree_join(branch, data[:location]['file']))

      data[:location]['blob_path'] = path
    end

    data
  end
end
