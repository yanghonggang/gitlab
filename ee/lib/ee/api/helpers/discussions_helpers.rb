# frozen_string_literal: true

module EE
  module API
    module Helpers
      module DiscussionsHelpers
        extend ActiveSupport::Concern

        class_methods do
          extend ::Gitlab::Utils::Override

          override :feature_category_per_noteable_type
          def feature_category_per_noteable_type
            super.merge!(::Epic => :issue_tracking)
          end
        end
      end
    end
  end
end
