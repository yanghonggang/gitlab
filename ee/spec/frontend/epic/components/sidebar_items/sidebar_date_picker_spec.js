import { GlLoadingIcon, GlIcon } from '@gitlab/ui';
import { shallowMount } from '@vue/test-utils';
import SidebarDatepicker from 'ee/epic/components/sidebar_items/sidebar_date_picker.vue';
import { TEST_HOST } from 'helpers/test_constants';
import DatePicker from '~/vue_shared/components/pikaday.vue';
import CollapsedCalendarIcon from '~/vue_shared/components/sidebar/collapsed_calendar_icon.vue';
import ToggleSidebar from '~/vue_shared/components/sidebar/toggle_sidebar.vue';
import { mockDatePickerProps } from '../../mock_data';

const mockPopoverBind = jest.fn();

jest.mock('~/vue_shared/directives/popover', () => ({
  bind: (...args) => mockPopoverBind(...args),
}));

describe('SidebarDatePicker', () => {
  let originalGon;
  beforeAll(() => {
    originalGon = global.gon;
    global.gon = { gitlab_url: TEST_HOST };
  });

  afterAll(() => {
    global.gon = originalGon;
  });

  let wrapper;

  const findIconByName = name =>
    wrapper
      .findAll(GlIcon)
      .filter(w => w.props().name === name)
      .at(0);
  const findEditButton = () => wrapper.find({ ref: 'editButton' });
  const findDirectiveCallByTitle = title =>
    mockPopoverBind.mock.calls.find(([, binding]) => binding.value.title === title);
  const findRemoveButton = () => wrapper.find({ ref: 'removeButton' });

  const createFakeEvent = () => ({ stopPropagation: jest.fn() });

  const startEditing = () => {
    const e = createFakeEvent();
    findEditButton().vm.$emit('click', e);
  };

  const createComponent = props => {
    wrapper = shallowMount(SidebarDatepicker, {
      propsData: {
        ...mockDatePickerProps,
        ...props,
      },
    });
  };

  afterEach(() => {
    wrapper.destroy();
    wrapper = null;
  });

  it('generates unique names for input if `fieldName` prop is not provided', () => {
    createComponent();

    const anotherWrapper = shallowMount(SidebarDatepicker, {
      propsData: mockDatePickerProps,
    });

    const firstInputName = wrapper.find('input').attributes('name');
    const otherInputName = anotherWrapper.find('input').attributes('name');
    expect(firstInputName).toContain('dateType_');
    expect(otherInputName).toContain('dateType_');
    expect(firstInputName).not.toEqual(otherInputName);
    anotherWrapper.destroy();
  });

  it('renders remove button when both `selectedDate` is defined and `canUpdate` is true', () => {
    createComponent({
      selectedDate: new Date(),
      dateFixed: new Date(),
      canUpdate: true,
    });

    expect(findRemoveButton().exists()).toBe(true);
  });

  describe('collapsed calendar icon', () => {
    it('receives full date string in words based on `selectedDate` prop value', () => {
      createComponent({
        selectedDate: new Date(2018, 0, 1),
      });

      expect(wrapper.find(CollapsedCalendarIcon).props('text')).toBe('Jan 1, 2018');
    });

    it('receives `None` when `selectedDateWords` is not defined', () => {
      createComponent();

      expect(wrapper.find(CollapsedCalendarIcon).props('text')).toBe('None');
    });
  });

  it('returns full date string in words based on `dateFixed` prop value', () => {
    createComponent({
      dateFixed: new Date(2018, 0, 1),
    });

    expect(wrapper.text()).toContain('Jan 1, 2018');
  });

  it('returns full date string in words when `dateFromMilestones` is defined', () => {
    createComponent({ dateFromMilestones: new Date(2018, 0, 1) });

    expect(wrapper.text()).toContain('Inherited: Jan 1, 2018');
  });

  it('returns `None` when `dateFromMilestones` is not defined', () => {
    createComponent();
    expect(wrapper.text()).toContain('Inherited: None');
  });

  it('passes correct popover options to directive', () => {
    createComponent();
    return wrapper.vm.$nextTick().then(() => {
      const expectedTitle =
        'These dates affect how your epics appear in the roadmap. Dates from milestones come from the milestones assigned to issues in the epic. You can also set fixed dates or remove them entirely.';
      const [, binding] = findDirectiveCallByTitle(expectedTitle);
      const { content, ...popoverConfig } = binding.value;
      delete popoverConfig.template;
      const expectedContent = '/help/user/group/epics/index.md#start-date-and-due-date';
      const expectedPopoverConfig = {
        html: true,
        trigger: 'focus',
        title: expectedTitle,

        container: 'body',
        boundary: 'viewport',
      };
      expect(mockPopoverBind).toHaveBeenCalled();

      expect(popoverConfig).toStrictEqual(expectedPopoverConfig);
      expect(content).toContain(expectedContent);
    });
  });

  it('returns popover config object containing title with appropriate string', () => {
    createComponent({ isDateInvalid: true, selectedDateIsFixed: false });
    return wrapper.vm.$nextTick().then(() => {
      const expectedTitle = 'Selected date is invalid';
      const [, targetBinding] = findDirectiveCallByTitle(expectedTitle);
      const { content } = targetBinding.value;

      expect(content).toContain('/help/user/group/epics/index.md#start-date-and-due-date');
      expect(content).toContain('How can I solve this?');
    });
  });

  it('stops editing and emits `toggleDateType` event on component on `hidePicker` from date picker', () => {
    createComponent({ canUpdate: true });
    startEditing();

    return wrapper.vm
      .$nextTick()
      .then(() => {
        wrapper.find(DatePicker).vm.$emit('hidePicker');
        expect(wrapper.emitted().toggleDateType[0]).toStrictEqual([true, true]);
      })
      .then(() => wrapper.vm.$nextTick())
      .then(() => {
        expect(wrapper.find(DatePicker).exists()).toBe(false);
      });
  });

  it('starts editing when clicked on edit button', () => {
    createComponent();
    expect(wrapper.find(DatePicker).exists()).toBe(false);

    const e = createFakeEvent();
    findEditButton().vm.$emit('click', e);

    return wrapper.vm.$nextTick().then(() => {
      expect(e.stopPropagation).toHaveBeenCalled();
      expect(wrapper.find(DatePicker).exists()).toBe(true);
    });
  });

  it('stops editing and emits `saveDate` when `newDateSelected` emitted by date picker', () => {
    const date = new Date();
    createComponent();
    startEditing();

    return wrapper.vm.$nextTick().then(() => {
      wrapper.find(DatePicker).vm.$emit('newDateSelected', date);
      expect(wrapper.emitted().saveDate).toStrictEqual([[date]]);
    });
  });

  it('emits `toggleDateType` event on component when input is clicked', () => {
    createComponent({ canUpdate: true });

    wrapper.find('input').trigger('click');

    return wrapper.vm.$nextTick(() => {
      expect(wrapper.emitted().toggleDateType).toStrictEqual([[true]]);
    });
  });

  it('emits `toggleCollapse` event when toggle-sidebar emits `toggle` event', () => {
    createComponent({ showToggleSidebar: true });
    wrapper.find(ToggleSidebar).vm.$emit('toggle');

    return wrapper.vm.$nextTick().then(() => {
      expect(wrapper.emitted().toggleCollapse).toBeDefined();
    });
  });

  it('renders expected template', () => {
    createComponent({
      fieldName: 'datetype_test',
    });

    expect(wrapper.element).toMatchSnapshot();
  });

  it('renders collapse button when `showToggleSidebar` prop is `true`', () => {
    createComponent({ showToggleSidebar: true });

    expect(wrapper.find(ToggleSidebar).exists()).toBe(true);
  });

  it('renders loading icon when `dateSaveInProgress` prop is true', () => {
    createComponent({ dateSaveInProgress: true });

    expect(wrapper.find(GlLoadingIcon).exists()).toBe(true);
  });

  it('renders date warning icon when `isDateInvalid` prop is `true`', () => {
    createComponent({ isDateInvalid: true, selectedDateIsFixed: false });
    const warningIcon = findIconByName('warning');

    expect(warningIcon.exists()).toBe(true);
    expect(warningIcon.attributes('tabindex')).toBe('0');
  });
});
