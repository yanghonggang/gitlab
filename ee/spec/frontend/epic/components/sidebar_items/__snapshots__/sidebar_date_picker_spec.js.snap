// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidebarDatePicker renders expected template 1`] = `
<div
  class="block date epic-date"
>
  <collapsed-calendar-icon-stub
    class="sidebar-collapsed-icon"
    containerclass=""
    showicon="true"
    text="None"
    tooltiptext=""
  />
   
  <div
    class="title"
  >
    
    Date
    
    <!---->
     
    <div
      class="float-right d-flex"
    >
      <gl-icon-stub
        class="help-icon gl-mr-2"
        name="question-o"
        size="16"
        tabindex="0"
      />
       
      <gl-button-stub
        buttontextclasses=""
        category="primary"
        class="btn-sidebar-action"
        icon=""
        size="medium"
        variant="link"
      >
        
        Edit
      
      </gl-button-stub>
       
      <!---->
    </div>
  </div>
   
  <div
    class="value"
  >
    <div
      class="value-type-fixed text-secondary is-option-selected d-flex"
    >
      <input
        name="datetype_test"
        type="radio"
      />
       
      <span
        class="gl-ml-2"
      >
        Fixed:
      </span>
       
      <span
        class="d-flex value-content gl-ml-1"
      >
        <span
          class="no-value"
        >
           None 
        </span>
      </span>
    </div>
     
    <abbr
      class="value-type-dynamic text-secondary d-flex gl-mt-3"
      title="Select an issue with milestone to set date"
    >
      <input
        name="datetype_test"
        type="radio"
      />
       
      <span
        class="gl-ml-2"
      >
        Inherited:
      </span>
       
      <span
        class="value-content gl-ml-1"
      >
        None
      </span>
       
      <!---->
    </abbr>
  </div>
</div>
`;
